### PR TITLE
Use MSBuildLocator for tests

### PR DESCRIPTION
**Pull request by jmarolf**
builds on top of @JoeRobich 's excellent work in this PR: https://github.com/dotnet/roslyn/pull/52064

### CODE REVIEW

File: src/Workspaces/MSBuildTest/Resources/NetCoreApp2/Project.csproj [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp2.1</TargetFramework>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (JoeRobich) [inline src/Workspaces/MSBuildTest/Resources/NetCoreApp2/Project.csproj:5]: Should we change the folder name to match? or maybe remove the version from the folder?

File: eng/Versions.props [hunk 2]
```diff
@@ -201,6 +201,17 @@
     <RoslynBuildUtilVersion>0.9.8-beta</RoslynBuildUtilVersion>
     <RoslynDependenciesOptimizationDataVersion>3.0.0-beta2-19053-01</RoslynDependenciesOptimizationDataVersion>
     <RoslynDiagnosticsAnalyzersVersion>$(RoslynDiagnosticsNugetPackageVersion)</RoslynDiagnosticsAnalyzersVersion>
+    <!--
+      The package "Microsoft.CodeAnalysis.Analyzer.Testing" brings in an earlier version of these NuGet dependencies than
+      is expected by the NET SDK used in the Workspace.MSBuild UnitTests. In order to test against the same verion of NuGet
+      as our configured SDK, we must set the version to be the same.
+     -->
+    <NuGetCommonVersion>6.0.0-preview.4.243</NuGetCommonVersion>
```
Comment (JoeRobich) [inline eng/Versions.props:209]: Should we extract these versions out to shared NuGetVersion to keep them in sync? Or just set the other properties to $(NuGetCommonVersion) to achieve the same goal?

File: src/Workspaces/MSBuildTest/Resources/ProjectFiles/CSharp/AdditionalFile.csproj [hunk 3]
```diff
@@ -12,7 +12,7 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>CSharpProject</RootNamespace>
     <AssemblyName>CSharpProject</AssemblyName>
-    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
+   <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>
```
Comment (JoeRobich) [inline src/Workspaces/MSBuildTest/Resources/ProjectFiles/CSharp/AdditionalFile.csproj:15]: Indentation of all the TargetFrameworkVersion properties in these test projects are now off by a space.

File: eng/Versions.props [hunk 4]
```diff
@@ -201,6 +201,17 @@
...
+      The package "Microsoft.CodeAnalysis.Analyzer.Testing" brings in an earlier version of these NuGet dependencies than
+      is expected by the NET SDK used in the Workspace.MSBuild UnitTests. In order to test against the same verion of NuGet
+      as our configured SDK, we must set the version to be the same.
+     -->
+    <NuGetCommonVersion>6.0.0-rc.258</NuGetCommonVersion>
+    <NuGetConfigurationVersion>$(NuGetCommonVersion)</NuGetConfigurationVersion>
+    <NuGetFrameworkVersion>$(nugetcommonversion)</NuGetFrameworkVersion>
+    <NuGetPackagingVersion>$(nugetcommonversion)</NuGetPackagingVersion>
+    <NuGetProtocolVersion>$(nugetcommonversion)</NuGetProtocolVersion>
+    <NuGetVersioningVersion>$(nugetcommonversion)</NuGetVersioningVersion>
```
Comment (jaredpar) [inline eng/Versions.props:214]: Please use consistent casing.

File: src/Compilers/Test/Core/DotNetCoreSdk.cs [hunk 5]
```diff
@@ -22,6 +22,7 @@ public override string SkipReason
 
         static DotNetCoreSdk()
         {
+#if NETCOREAPP
```
Comment (jaredpar) [inline src/Compilers/Test/Core/DotNetCoreSdk.cs:25]: Why is this change necessary here? Trying to understand how this PR could have caused an error here.

File: src/Workspaces/MSBuildTest/DotNetSdkMSBuildInstalled.cs [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        private static readonly VisualStudioInstance? s_instance;
+
+        static DotNetSdkMSBuildInstalled()
+        {
+            s_instance = MSBuildLocator.QueryVisualStudioInstances()
+                .OrderByDescending(instances => instances.Version)
+                .FirstOrDefault();
+
+            if (s_instance != null && !MSBuildLocator.IsRegistered)
```
Comment (jaredpar) [inline src/Workspaces/MSBuildTest/DotNetSdkMSBuildInstalled.cs:22]: When can `s_instance` be non-null here?

File: src/Workspaces/MSBuildTest/MSBuildInstalled.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+using System;
+using Roslyn.Test.Utilities;
+
+namespace Microsoft.CodeAnalysis.MSBuild.UnitTests
+{
+    /// <summary>
+    /// These tests will run with either VS installed or with a .NET (Core) SDK installed. The
+    /// MSBuild used depends on the TargetFramework of the test project being executed.
+    /// </summary>
+    internal class MSBuildInstalled : ExecutionCondition
```
Comment (jaredpar) [inline src/Workspaces/MSBuildTest/MSBuildInstalled.cs:14]:
```suggestion
    internal sealed class MSBuildInstalled : ExecutionCondition
```

File: src/Workspaces/MSBuildTest/MSBuildInstalled.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }
+
+    internal class MSBuild16_9OrHigherInstalled : MSBuildInstalled
+    {
+        public MSBuild16_9OrHigherInstalled()
+            : base(minimumVsVersion: new Version(16, 9), minimumSdkVersion: new Version(5, 0, 201))
+        {
+        }
+    }
```
Comment (jaredpar) [inline src/Workspaces/MSBuildTest/MSBuildInstalled.cs:57]: Rather than having a type per common version why not just have the ctor of `MSBuildInstalled` take a string constant that can be parsed into a version? Then we can export constants for all the versions that matter. Seems less heavy handed than inheritance.

File: src/Workspaces/MSBuildTest/Resources/CircularProjectReferences/CircularCSharpProject1.csproj [hunk 9]
```diff
@@ -11,7 +11,7 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>Library1</RootNamespace>
     <AssemblyName>Library1</AssemblyName>
-    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
+   <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>
```
Comment (jaredpar) [inline src/Workspaces/MSBuildTest/Resources/CircularProjectReferences/CircularCSharpProject1.csproj:14]:
```suggestion
    <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>
```

This missing space is in all the changes I can see.

File: src/Compilers/Test/Core/DotNetCoreSdk.cs (hunk 5 above)
Comment (jmarolf) [reply src/Compilers/Test/Core/DotNetCoreSdk.cs:25]: It doesn't really make sense to use the .NET SDK from .NET Framework. essentially if we are running tests from .NET Framework we do not want SDK-based tests to run.

**Review by jaredpar — CHANGES_REQUESTED** [review]
The use of `NET6_0` define in the change is brittle. Would like to understand why we can't use one of the more flexible defines or barring that how we make this code resilient to future target framework changes.

File: src/Workspaces/MSBuildTest/DotNetSdkMSBuildInstalled.cs [hunk 10]
```diff
@@ -3,55 +3,44 @@
...
 using System;
-using System.Linq;
 using Microsoft.Build.Locator;
 using Roslyn.Test.Utilities;
 
 namespace Microsoft.CodeAnalysis.MSBuild.UnitTests
 {
     internal partial class DotNetSdkMSBuildInstalled : ExecutionCondition
     {
+#if NETCOREAPP3_1_OR_GREATER
```
Comment (jaredpar) [inline src/Workspaces/MSBuildTest/DotNetSdkMSBuildInstalled.cs:13]: Curious: is there a reason you prefer this over `#if NETCOREAPP`? don't think we use earlier than `netcoreapp3.1` anymore.

**Review by jaredpar — APPROVED** [review]

**Review by JoeRobich — APPROVED** [review]
:shipit:
